src/policy/fees.cpp

```diff
@@ -604,6 +604,16 @@ bool CBlockPolicyEstimator::processBlockTx(unsigned int nBlockHeight, const CTxM
         return false;
     }
 
+    // Do not consider this transaction's feerate if it might have been mined due to one of
+    // its children (CPFP).
+    // Note this check is only best-effort, as the child might not be part of the block. Or
+    // another child might be part of the block that we didn't have in our mempool.
+    CFeeRate feerate_entry{entry->GetFee(), (uint32_t)entry->GetTxSize()};
+    for (const CTxMemPoolEntry& child : entry->GetMemPoolChildren()) {
+        CFeeRate ancestor_score{child.GetModFeesWithAncestors(), (uint32_t)child.GetSizeWithAncestors()};
+        if (ancestor_score > feerate_entry) return false;
```

Comment:
Hm, I don't think `ancestor_score > individual_feerate` is really an indication that it was CPFP'd. We'd want to see if it had descendants in the block. It could just mean that this transaction had high-feerate ancestors.

Suggested alternative: I think `ancestor_score < individual_feerate` would be a good indication that this tx was the sponsor of a CPFP in the block. Maybe we use that condition, and drop all of the transaction's ancestors from the fee estimator?  Or, to reduce over-pruning, we drop the ancestors whose feerates are lower than this transaction's?

Reply:
(TL;DR your suggestion is best but i'm leaving the response to the first section for context)

Yeah it's a best effort. I could make sure that the descendant was also part of the block. Is it worth it though? If an ancestor-less transaction with mempool child that pay a higher feerate than itself got mined, the child probably got included as well (not necessarily, but i assumed very probably).
Do you think that assumption isn't reasonable? My doubt is more with more convoluted packages. For instance:
```
A ------ D
      __/ /
B /      /
      _ /
C /
```
With `A` 1sat/vb, `B` 10sat/vb, `C` 10sat/vb and `D` 1sat/vb. It would be a false positive for `A`. To prevent this we could simply condition the check on the descendants feerate of `A`:
```patch
diff --git a/src/policy/fees.cpp b/src/policy/fees.cpp
index 029dd108c3..5d5dfb2efe 100644
--- a/src/policy/fees.cpp
+++ b/src/policy/fees.cpp
@@ -609,9 +609,12 @@ bool CBlockPolicyEstimator::processBlockTx(unsigned int nBlockHeight, const CTxM
     // Note this check is only best-effort, as the child might not be part of the block. Or
     // another child might be part of the block that we didn't have in our mempool.
     CFeeRate feerate_entry{entry->GetFee(), (uint32_t)entry->GetTxSize()};
-    for (const CTxMemPoolEntry& child : entry->GetMemPoolChildren()) {
-        CFeeRate ancestor_score{child.GetModFeesWithAncestors(), (uint32_t)child.GetSizeWithAncestors()};
-        if (ancestor_score > feerate_entry) return false;
+    CFeeRate feerate_desc{entry->GetModFeesWithDescendants(), (uint32_t)entry->GetSizeWithDescendants()};
+    if (feerate_desc > feerate_entry) {
+        for (const CTxMemPoolEntry& child : entry->GetMemPoolChildren()) {
+            CFeeRate ancestor_score{child.GetModFeesWithAncestors(), (uint32_t)child.GetSizeWithAncestors()};
+            if (ancestor_score > feerate_entry) return false;
+        }
     }
 
     // How many blocks did it take for miners to include this transaction?
```

----

Regarding your suggestion. do you mean we should iterate over all the mempool entries that were part of the block to detect CPFPs beforehand? I agree it's much better. It makes sure that the child was part of the same block. And your suggestion to only drop the ancestors whose feerates are lower than this transaction's covers the case i detailed above. Nice! Is that roughly what you were thinking about:
```patch
diff --git a/src/policy/fees.cpp b/src/policy/fees.cpp
index 029dd108c3..8c0e71ea18 100644
--- a/src/policy/fees.cpp
+++ b/src/policy/fees.cpp
@@ -604,16 +604,6 @@ bool CBlockPolicyEstimator::processBlockTx(unsigned int nBlockHeight, const CTxM
         return false;
     }
 
-    // Do not consider this transaction's feerate if it might have been mined due to one of
-    // its children (CPFP).
-    // Note this check is only best-effort, as the child might not be part of the block. Or
-    // another child might be part of the block that we didn't have in our mempool.
-    CFeeRate feerate_entry{entry->GetFee(), (uint32_t)entry->GetTxSize()};
-    for (const CTxMemPoolEntry& child : entry->GetMemPoolChildren()) {
-        CFeeRate ancestor_score{child.GetModFeesWithAncestors(), (uint32_t)child.GetSizeWithAncestors()};
-        if (ancestor_score > feerate_entry) return false;
-    }
-
     // How many blocks did it take for miners to include this transaction?
     // blocksToConfirm is 1-based, so a transaction included in the earliest
     // possible block has confirmation count of 1
@@ -625,6 +615,7 @@ bool CBlockPolicyEstimator::processBlockTx(unsigned int nBlockHeight, const CTxM
         return false;
     }
 
+    CFeeRate feerate_entry{entry->GetFee(), (uint32_t)entry->GetTxSize()};
     feeStats->Record(blocksToConfirm, (double)feerate_entry.GetFeePerK());
     shortStats->Record(blocksToConfirm, (double)feerate_entry.GetFeePerK());
     longStats->Record(blocksToConfirm, (double)feerate_entry.GetFeePerK());
@@ -659,6 +650,18 @@ void CBlockPolicyEstimator::processBlock(unsigned int nBlockHeight,
     shortStats->UpdateMovingAverages();
     longStats->UpdateMovingAverages();
 
+    for (const auto& entry : entries) {
+        CFeeRate individual_feerate{entry->GetFee(), (uint32_t)entry->GetTxSize()};
+        CFeeRate ancestor_score{entry->GetModFeesWithAncestors(), (uint32_t)entry->GetSizeWithAncestors()};
+
+        if (individual_feerate > ancestor_score) {
+            for (const CTxMemPoolEntry& parent : entry->GetMemPoolParents()) {
+                CFeeRate parent_feerate{parent.GetFee(), (uint32_t)parent.GetTxSize()};
+                if (parent_feerate < individual_feerate) _removeTx(parent.GetTx().GetHash(), true);
+            }
+        }
+    }
+
     unsigned int countedTxs = 0;
     // Update averages with data points from current block
     for (const auto& entry : entries) {
```